runtime/src/epoch_rewards_hasher.rs

```diff
@@ -0,0 +1,156 @@
+use {
+    crate::bank::StakeRewards,
+    siphasher::sip::SipHasher13,
+    solana_sdk::{hash::Hash, pubkey::Pubkey},
+    std::hash::Hasher,
+};
+
+#[derive(Debug, Clone, Copy)]
+pub(crate) struct EpochRewardHasher {
+    hasher: SipHasher13,
+    partitions: usize,
+}
+
+impl EpochRewardHasher {
+    /// Use SipHasher13 keyed on the `seed` for calculating epoch reward partition
+    pub(crate) fn new(partitions: usize, hash: &Hash) -> Self {
+        let mut hasher = SipHasher13::new();
+        hasher.write(hash.as_ref());
+        Self { hasher, partitions }
+    }
+
+    /// Return partition index (0..partitions) by hashing `address` with the `hasher`
+    pub(crate) fn hash_address_to_partition(self, address: &Pubkey) -> usize {
+        let Self {
+            mut hasher,
+            partitions,
+        } = self;
+        hasher.write(address.as_ref());
+        let hash64 = hasher.finish();
+
+        hash_to_partition(hash64, partitions)
+    }
+}
+
+/// Compute the partition index by modulo the address hash to number of partitions w.o bias.
+/// (rand_int * DESIRED_RANGE_MAX) / (RAND_MAX + 1)
+fn hash_to_partition(hash: u64, partitions: usize) -> usize {
+    ((partitions as u128)
+        .saturating_mul(u128::from(hash))
+        .saturating_div(u128::from(u64::MAX).saturating_add(1))) as usize
+}
+
+#[allow(dead_code)]
+fn hash_rewards_into_partitions(
+    stake_rewards: &StakeRewards,
+    parent_block_hash: &Hash,
+    num_partitions: usize,
+) -> Vec<StakeRewards> {
+    let hasher = EpochRewardHasher::new(num_partitions, parent_block_hash);
+    let mut result = vec![vec![]; num_partitions];
+
+    for reward in stake_rewards {
+        let partition_index = hasher.hash_address_to_partition(&reward.stake_pubkey);
```

Comment:
80.

Reply:
I reworked this to just copy the hasher itself each time we ask for a new hash. The assumption is the copy of the hasher and its state is faster than rehashing the seed each time.

Reply:
> I reworked this to just copy the hasher itself each time we ask for a new hash. The assumption is the copy of the hasher and its state is faster than rehashing the seed each time.

to what end? i think it's actually confusing now because we have an associated function masquerading as a method. afaict brooks' concern was more a "hey are you sure you want to be copying these?" to which the answer is "yes".

imo the "fix" is a comment and/or removing the `Copy` derive from `EpochRewardHasher` and explicitly `hasher().clone().hash_addre...` here

Reply:
I can make it a fn that is not in the impl if that helps. If we put a clone() on the callsite, then everyone who comes along after will ask if that is necessary or intended. A comment can be out of date and also means the impl could be better such that we don't need a comment.
I definitely don't like the implicit copy.
As a minor detail, it isn't necessary to clone/copy the num_partitions as well since that stays the same. We just need to copy the hash state. Then, we never need that copied state again after we finish the hash right now.

Reply:
> I can make it a fn that is not in the impl if that helps.

i'm still not sure what exactly we're trying to fix

> If we put a clone() on the callsite, then everyone who comes along after will ask if that is necessary or intended.

but paired with a doc comment, that should be self-resolving

> A comment can be out of date and also means the impl could be better such that we don't need a comment.

again, i don't know what we're trying to fix, so i can't make a claim as to "better"

> I definitely don't like the implicit copy.

:handshake:

> As a minor detail, it isn't necessary to clone/copy the num_partitions as well since that stays the same. We just need to copy the hash state.

yeah but the hash state is 90% of the copy load and we're well within two cache lines so probably better to err on the side of maintainability until benchmarking actually raises this as a problem

> Then, we never need that copied state again after we finish the hash right now.

which is why the method was originally changed to consume the object

Reply:
The problem we're trying to solve is that it ideally doesn't need a comment.

But I will change it back to take self by value, keep the removal of the copy attribute so there is no implicit copy, make an explicit clone at the callsite, with a comment saying why we are cloning it.

Reply:
pushed a change to go back to taking by value and an explicit clone